cloud_function/cloud_function_rules.bzl

```diff
@@ -0,0 +1,222 @@
+    )
+
+    out = depset(direct = [f])
+    return [
+        DefaultInfo(
+            files = out,
+        ),
+        OutputGroupInfo(
+            all_files = out,
+        ),
+    ]
+
+_py_lambda_zip = rule(
+    implementation = _py_lambda_zip_impl,
+    attrs = {
+        "target": attr.label(),
+        "ignore": attr.string_list(),
+        "_zipper": attr.label(
+            default = Label("@bazel_tools//tools/zip:zipper"),
+            cfg = "host",
+            executable = True,
+        ),
+        "output": attr.output(),
+    },
+    executable = False,
+    test = False,
+)
+
+def py_lambda_zip(name, target, ignore, **kwargs):
+    _py_lambda_zip(
+        name = name,
+        target = target,
+        ignore = ignore,
+        output = name + ".zip",
+        **kwargs
+    )
+
+# Just a wrapper of gen rule:
+# genrule(
+#     name = "upload_to_s3",
+#     srcs = [":lambda_archive"],
+#     outs = ["upload_to_s3_done"],
+#     cmd = """
+#         CURRENT_DATE=`date '+%Y/%m/%d'`
+#         FILE_NAME=`cd services/ && git rev-parse HEAD`.zip
+#         aws --profile prod-engineering-labs s3 cp $(location :lambda_archive) s3://6si-dheeraj-test-tmp/$$CURRENT_DATE/$$FILE_NAME &&
+#         touch $@
+#     """,
+# )
+
+def _upload_to_s3_impl(ctx):
+    output = ctx.actions.declare_file(ctx.attr.name + "_done")
+
+    profile_option = ""
+    if ctx.attr.profile != "":
+        profile_option = "--profile " + ctx.attr.profile
+
+    key_prefix_option = ""
+    if ctx.attr.key_prefix != "":
+        key_prefix_option = ctx.attr.key_prefix + "/"
+
+    cmd = """
+        set -e
+        CURRENT_DATE=`date '+%Y/%m/%d'`
+        FILE_NAME=`cd services/ && git rev-parse HEAD`.zip
+        KEY={key_prefix}{function_name}/$CURRENT_DATE/$FILE_NAME
+        aws {profile_option} s3 cp {src} s3://{bucket}/$KEY
+        echo S3_SOURCE > {out}
+        echo $KEY >> {out}
+        echo {function_name} >> {out}
+        echo {bucket} >> {out}
+    """.format(
+        src=ctx.file.src.path, 
+        out=output.path, 
+        function_name=ctx.attr.function_name,
+        profile_option=profile_option, 
+        bucket=ctx.attr.bucket,
+        key_prefix=key_prefix_option,
+    )
+
+    ctx.actions.run_shell(
+        outputs=[output],
+        inputs=[ctx.file.src],
+        command=cmd,
+    )
+
+    return DefaultInfo(files=depset([output]))
+
+
+lambda_upload_s3 = rule(
+    implementation=_upload_to_s3_impl,
+    attrs={
+        "src": attr.label(allow_single_file=True),
+        # function_name MUST MATCH the name of the lambda function within AWS
+        "function_name": attr.string(),
+        # AWS config profile to use
+        "profile": attr.string(default=""),
+        # S3 Bucket
+        "bucket": attr.string(default="6si-lambda"),
+        # S3 key Prefix, useful for testing
+        "key_prefix": attr.string(default=""),
+    },
+    outputs={"done": "%{name}_done"},
+)
+
+
+def _lambda_deploy_code_imp(ctx):
```

Comment:
Sorry for the late comment, but isn't the deployment handled by Pulumi?

Reply:
yes it will be, that is not yet ready however, so in the meantime we will use this function